<img src="http://imgur.com/1ZcRyrc.png" style="float: left; margin: 20px; height: 55px">

## Lab: Cleaning Rock Song Data

_Authors: Dave Yerrington (SF)_

---


In [1]:
import pandas as pd
import numpy as np 
import seaborn as sns

%matplotlib inline

### 1. Load `rock.csv` and do an initial examination of its data columns.

In [2]:
rockfile = "./datasets/rock.csv"

In [125]:
df = pd.read_csv(rockfile)

In [127]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2230 entries, 0 to 2229
Data columns (total 8 columns):
Song Clean      2230 non-null object
ARTIST CLEAN    2230 non-null object
Release Year    1653 non-null object
COMBINED        2230 non-null object
First?          2230 non-null int64
Year?           2230 non-null int64
PlayCount       2230 non-null int64
F*G             2230 non-null int64
dtypes: int64(4), object(4)
memory usage: 139.4+ KB


In [5]:
df.describe()

First?        Year?    PlayCount          F*G
count  2230.0  2230.000000  2230.000000  2230.000000
mean      1.0     0.741256    16.872646    15.048430
std       0.0     0.438043    25.302972    25.288366
min       1.0     0.000000     0.000000     0.000000
25%       1.0     0.000000     1.000000     0.000000
50%       1.0     1.000000     4.000000     3.000000
75%       1.0     1.000000    21.000000    18.000000
max       1.0     1.000000   142.000000   142.000000

In [6]:
df.head(3)

Song Clean ARTIST CLEAN Release Year  \
0  Caught Up in You  .38 Special         1982   
1      Fantasy Girl  .38 Special          NaN   
2   Hold On Loosely  .38 Special         1981   

                          COMBINED  First?  Year?  PlayCount  F*G  
0  Caught Up in You by .38 Special       1      1         82   82  
1      Fantasy Girl by .38 Special       1      0          3    0  
2   Hold On Loosely by .38 Special       1      1         85   85

### 2.  Clean up the column names.

Let's clean up the column names. There are two ways we can accomplish this:

#### 2.A Change the column names when you import the data using `pd.read_csv()`.

Notice that, when passing `names=[..A LIST OF STRING..]` with a number of columns that matches the number of strings in the passed list, you replace the column names.

NOTE: When you create custom column names, the first row of the `.csv` already represents a header. It is important to tell `pandas` to skip that row. The `skiprows=1` keyword argument to `read_csv()` will tell `pandas` to skip the first row.

In [7]:
import re
newcolnames = []
for c in df.columns:
    newcolnames.append(re.sub(r'\W+','',c))

newcolnames

['SongClean',
 'ARTISTCLEAN',
 'ReleaseYear',
 'COMBINED',
 'First',
 'Year',
 'PlayCount',
 'FG']

In [34]:
df2 = pd.read_csv(rockfile, names=newcolnames, skiprows=1)
test = df2

In [33]:
df2.head()

SongClean  ARTISTCLEAN ReleaseYear  \
0        Caught Up in You  .38 Special        1982   
1            Fantasy Girl  .38 Special         NaN   
2         Hold On Loosely  .38 Special        1981   
3  Rockin' Into the Night  .38 Special        1980   
4       Art For Arts Sake         10cc        1975   

                                COMBINED  First  Year  PlayCount  FG  
0        Caught Up in You by .38 Special      1     1         82  82  
1            Fantasy Girl by .38 Special      1     0          3   0  
2         Hold On Loosely by .38 Special      1     1         85  85  
3  Rockin' Into the Night by .38 Special      1     1         18  18  
4              Art For Arts Sake by 10cc      1     1          1   1

#### 2.B Change column names using the `.rename()` function.

The `.rename()` function takes an argument, `columns=name_dict`, in which `name_dict` is a dictionary containing the original column names as keys and the new column names as values.

In [10]:
newcolnames_dict = dict(zip(df.columns, newcolnames))



df.rename(columns = newcolnames_dict, inplace=True)


In [11]:
#rename specific column
df.rename(columns={'ARTISTCLEAN':'ArtistClean'}, inplace=True)
df.head(3)

SongClean  ArtistClean ReleaseYear                         COMBINED  \
0  Caught Up in You  .38 Special        1982  Caught Up in You by .38 Special   
1      Fantasy Girl  .38 Special         NaN      Fantasy Girl by .38 Special   
2   Hold On Loosely  .38 Special        1981   Hold On Loosely by .38 Special   

   First  Year  PlayCount  FG  
0      1     1         82  82  
1      1     0          3   0  
2      1     1         85  85

#### 2.C Reassigning the `.columns` attribute of a DataFrame.

You can also just reassign the `.columns` attribute to a list of strings containing the new column names. 

The only caveat with reassigning `.columns` is that you have to reassign all of the column names at once. You can't partially replace a value by working on `.columns` directly. You have to reassign `.columns` with a list of equal length. 

In [134]:
df.columns = newcolnames
df.head(3)

SongClean  ARTISTCLEAN ReleaseYear                         COMBINED  \
0  Caught Up in You  .38 Special        1982  Caught Up in You by .38 Special   
1      Fantasy Girl  .38 Special         NaN      Fantasy Girl by .38 Special   
2   Hold On Loosely  .38 Special        1981   Hold On Loosely by .38 Special   

   First  Year  PlayCount  FG  
0      1     1         82  82  
1      1     0          3   0  
2      1     1         85  85

### 3. Subsetting data where null values exist.

We have mixed `str` and `NaN` values in the `release` column. `NaN` stands for "not a number" and is the way `pandas` handles "nulls" or nonexistent data. We can use the `.isnull()` method of a Series to find null values.

Print the header of the data subset to where the `release` column is null values.

In [135]:
df4 = df['ReleaseYear'].isnull()

In [136]:
df4.columns = ['ReleaseYear']
df4.head(3)

0    False
1     True
2    False
Name: ReleaseYear, dtype: bool

In [137]:
df['ReleaseYear'].isnull().sum()

577

In [140]:
df[df['ReleaseYear'].isnull()].head()

SongClean  ARTISTCLEAN ReleaseYear  \
1             Fantasy Girl  .38 Special         NaN   
10   Baby, Please Don't Go        AC/DC         NaN   
13  CAN'T STOP ROCK'N'ROLL        AC/DC         NaN   
16        Girls Got Rhythm        AC/DC         NaN   
24         Let's Get It Up        AC/DC         NaN   

                           COMBINED  First  Year  PlayCount  FG  
1       Fantasy Girl by .38 Special      1     0          3   0  
10   Baby, Please Don't Go by AC/DC      1     0          1   0  
13  CAN'T STOP ROCK'N'ROLL by AC/DC      1     0          5   0  
16        Girls Got Rhythm by AC/DC      1     0         24   0  
24         Let's Get It Up by AC/DC      1     0          4   0

In [141]:
df[~df['ReleaseYear'].isnull()].head()

SongClean   ARTISTCLEAN ReleaseYear  \
0        Caught Up in You   .38 Special        1982   
2         Hold On Loosely   .38 Special        1981   
3  Rockin' Into the Night   .38 Special        1980   
4       Art For Arts Sake          10cc        1975   
5              Kryptonite  3 Doors Down        2000   

                                COMBINED  First  Year  PlayCount  FG  
0        Caught Up in You by .38 Special      1     1         82  82  
2         Hold On Loosely by .38 Special      1     1         85  85  
3  Rockin' Into the Night by .38 Special      1     1         18  18  
4              Art For Arts Sake by 10cc      1     1          1   1  
5             Kryptonite by 3 Doors Down      1     1         13  13

### 4. Update slices of your DataFrame based on mask selection/slices.

In many scenarios, we want to upate values in our DataFrame according to criteria. Let's say we wanted to set all of the null values in `release` to 0.

With newer versions of `pandas`, in order to manipulate data in the original DataFrame, we have to use `.loc` while performing reassignment using a mask and an index.

For example, the following won't always work:
```python
df[row_mask]['column_name'] = new_value
```

The best way to accomplish the same task is:
```python
df.loc[row_mask, 'column_name'] = new_value
```

For multiple column assignment, you would use:
```python
df.loc[row_mask, ['col_1', 'col_2', 'col_3']] = new_value
```

#### 4.A Let's try it out. Make all of the null values in `release` 0.

In [64]:
df.loc[df['ReleaseYear'].isnull(),'ReleaseYear'] = 0

In [65]:
df.head(3)

SongClean  ARTISTCLEAN  ReleaseYear  \
0  Caught Up in You  .38 Special       1982.0   
1      Fantasy Girl  .38 Special          0.0   
2   Hold On Loosely  .38 Special       1981.0   

                          COMBINED  First  Year  PlayCount  FG  True  
0  Caught Up in You by .38 Special      1     1         82  82   NaN  
1      Fantasy Girl by .38 Special      1     0          3   0   NaN  
2   Hold On Loosely by .38 Special      1     1         85  85   NaN

#### 4.B Verify that `release` contains no null values.

In [66]:
df['ReleaseYear'].isnull().sum()

0

### 5. Ensure that the data types of the columns make sense. 

Verifying column data types is a critical part of data munging. If columns have the wrong data type, then there is usually corrupted or incorrect data in some of the observations.

#### 5.A Look at the data types for the columns. Are any incorrect given what the data represents?

In [67]:
df.dtypes

SongClean       object
ARTISTCLEAN     object
ReleaseYear    float64
COMBINED        object
First            int64
Year             int64
PlayCount        int64
FG               int64
True           float64
dtype: object

### 6. Investigate and clean up the `release` column.

The `release` column is a string data type when it should be an integer.

#### 6.A Figure out what value(s) are causing the `release` column to be encoded as a string instead of an integer.

In [142]:
for c in df['ReleaseYear']:
    try:
        float(c)
    except:
        print(c)

SONGFACTS.COM


In [68]:
df['ReleaseYear'].value_counts()

0.0       578
1973.0    104
1975.0     83
1977.0     83
1970.0     81
1971.0     75
1969.0     72
1980.0     70
1978.0     64
1979.0     63
1981.0     61
1967.0     61
1983.0     60
1976.0     56
1982.0     54
1984.0     51
1972.0     50
1974.0     48
1968.0     46
1985.0     39
1987.0     39
1986.0     37
1991.0     34
1989.0     32
1966.0     30
1988.0     29
1965.0     28
1994.0     25
1990.0     22
1993.0     19
1992.0     14
1964.0     14
1999.0     13
1995.0     10
1997.0      9
1963.0      9
1996.0      9
1998.0      6
2002.0      6
2005.0      5
2012.0      5
2004.0      5
2001.0      4
1962.0      3
2000.0      3
2007.0      3
2003.0      3
2011.0      3
2008.0      3
2013.0      2
2014.0      2
1961.0      1
2006.0      1
1958.0      1
1955.0      1
1071.0      1
Name: ReleaseYear, dtype: int64

#### 6.B Look at the rows in which there is incorrect data in the `release` column.

In [22]:
df[df['ReleaseYear']=='SONGFACTS.COM']

SongClean     ARTISTCLEAN    ReleaseYear  \
1504  Bullfrog Blues  Rory Gallagher  SONGFACTS.COM   

                              COMBINED  First  Year  PlayCount  FG  
1504  Bullfrog Blues by Rory Gallagher      1     1          1   1

#### 6.C. Clean up the data. Normally we may replace the offending data with null np.nan values, however we previously converted all of the nan values in the release column to zeros so we might as well continue with the same practice. Replacing with 0 (or nan) will allow us to convert the column to numeric.

In [145]:
df.loc[df['ReleaseYear'] == 'SONGFACTS.COM','ReleaseYear'] = 0
df['ReleaseYear'] = pd.to_numeric(df['ReleaseYear'])
df.dtypes

SongClean       object
ARTISTCLEAN     object
ReleaseYear    float64
COMBINED        object
First            int64
Year             int64
PlayCount        int64
FG               int64
dtype: object

### 7. Get summary statistics for the `release` column using the `.describe()` function.

Now that the `release` column is finally a numeric data type, we can apply the `.describe()` function.  

#### 7.A Print out the summary stats for the `release` column. What is the earliest and latest release date?

In [146]:
df['ReleaseYear'].describe()

count    1653.000000
mean     1976.823351
std        54.330553
min         0.000000
25%      1971.000000
50%      1977.000000
75%      1984.000000
max      2014.000000
Name: ReleaseYear, dtype: float64

In [148]:
df[df['ReleaseYear']>0]['ReleaseYear'].describe()

count    1652.000000
mean     1978.019976
std        24.191247
min      1071.000000
25%      1971.000000
50%      1977.000000
75%      1984.000000
max      2014.000000
Name: ReleaseYear, dtype: float64

#### 7.B Based on the summary statistics, is there anything else wrong with the `release` column? 

In [25]:
df['ReleaseYear'].unique()

array([1982.,   nan, 1981., 1980., 1975., 2000., 2002., 1992., 1985.,
       1993., 1976., 1995., 1979., 1984., 1977., 1990., 1986., 1974.,
       2014., 1987., 1973., 2001., 1989., 1997., 1971., 1972., 1994.,
       1970., 1966., 1965., 1983., 1955., 1978., 1969., 1999., 1968.,
       1988., 1962., 2007., 1967., 1958., 1071., 1996., 1991., 2005.,
       2011., 2004., 2012., 2003., 1998., 2008., 1964., 2013., 2006.,
          0., 1963., 1961.])

_Looking at the DataFrame that contains the year 1071, we can see that the year was probably corrupted and should be replaced with something else if possible._

### 8. Make changes and investigate using custom functions with `.apply()`.

Let's say we want to traverse every single row in our data set and apply a function to that row.

#### 8.A Write a function that will take a row of a DataFrame and print out the song, artist, and whether or not the release date is < 1970.


In [85]:
df['<1970'] = df['ReleaseYear'].apply(lambda x: x<1970)

In [104]:
def release_date(row):
    if row['ReleaseYear'] < 1970:
        print('Song: {}, Artist: {}, Release<{}: {}'.format(row['SongClean'],row['ARTISTCLEAN'],1970,True))
    else:
        print('Song: {}, Artist: {}, Release<{}: {}'.format(row['SongClean'],row['ARTISTCLEAN'],1970,False))

In [105]:
release_date(df.iloc[2])

Song: Hold On Loosely, Artist: .38 Special, Release<1970: False


#### 8.B Using the `.apply()` function, apply the function you wrote to the first four rows of the DataFrame.

You will need to tell the `apply` function to operate row by row. Setting the keyword argument as `axis=1` indicates that the function should be applied to each row individually.

In [172]:
df.loc[[1,2,5],['SongClean','ReleaseYear']]

SongClean  ReleaseYear
1     Fantasy Girl          NaN
2  Hold On Loosely       1981.0
5       Kryptonite       2000.0

In [155]:
df[:4].apply(release_date, axis=1)

Song: Caught Up in You, Artist: .38 Special, Release<1970: False
Song: Fantasy Girl, Artist: .38 Special, Release<1970: False
Song: Hold On Loosely, Artist: .38 Special, Release<1970: False
Song: Rockin' Into the Night, Artist: .38 Special, Release<1970: False


0    None
1    None
2    None
3    None
dtype: object

You'll notice that there will be a final output Series of `None` values. The `.apply()` function, if a return value is not specified, will return a Series of `None` values (similar to how the default return for Python functions is `None` when a return statement is not specified).

### 9. Write a function that converts cells in a DataFrame to float and otherwise replaces them with `np.nan`.

If applied to our data, it would keep only the numeric information and otherwise input null values.

Recall that the try-except syntax in Python is a great way to try something and take another action if the initial step fails:

```python
try:
    Perform some action.
except:
   Perform some other action if the first failed with an error.
```

#### 9.A Write the function that takes a column and converts all of its values to float if possible and `np.nan` otherwise. The return value should be the converted Series.

In [52]:
def col_convert(column):
    cleanls = []
    for cell in column:
        try:
            cleanls.append(float(cell))
        except:
            cleanls.append(np.nan)
    
    cleanseries = pd.Series(cleanls)
    return cleanseries

In [108]:
col_convert(test['ReleaseYear'])

0       1982.0
1          NaN
2       1981.0
3       1980.0
4       1975.0
5       2000.0
6       2000.0
7       2002.0
8       1992.0
9       1985.0
10         NaN
11      1980.0
12      1993.0
13         NaN
14      1976.0
15      1981.0
16         NaN
17      1995.0
18      1980.0
19      1980.0
20      1979.0
21      1975.0
22      1984.0
23      1977.0
24         NaN
25         NaN
26         NaN
27      1979.0
28      1980.0
29         NaN
         ...  
2200    1983.0
2201    1971.0
2202    1971.0
2203       NaN
2204    1971.0
2205       NaN
2206       NaN
2207    1968.0
2208       NaN
2209    1976.0
2210    1973.0
2211    1980.0
2212    1983.0
2213    1983.0
2214    1975.0
2215       NaN
2216       NaN
2217    1973.0
2218       NaN
2219    1973.0
2220    1983.0
2221       NaN
2222       NaN
2223    1981.0
2224    1983.0
2225       NaN
2226    1981.0
2227    1975.0
2228    1983.0
2229    1973.0
Length: 2230, dtype: float64

In [107]:
clean_series = col_convert(test['ReleaseYear'])

In [109]:
df.columns

Index([u'SongClean', u'ARTISTCLEAN', u'ReleaseYear', u'COMBINED', u'First',
       u'Year', u'PlayCount', u'FG', u'True', u'<1970'],
      dtype='object')

#### 9.B Try your function out on the rock song data and ensure the output is what you expected.


In [114]:
df.head()

SongClean  ARTISTCLEAN  ReleaseYear  \
0        Caught Up in You  .38 Special       1982.0   
1            Fantasy Girl  .38 Special          0.0   
2         Hold On Loosely  .38 Special       1981.0   
3  Rockin' Into the Night  .38 Special       1980.0   
4       Art For Arts Sake         10cc       1975.0   

                                COMBINED  First  Year  PlayCount  FG  True  \
0        Caught Up in You by .38 Special      1     1         82  82   NaN   
1            Fantasy Girl by .38 Special      1     0          3   0   NaN   
2         Hold On Loosely by .38 Special      1     1         85  85   NaN   
3  Rockin' Into the Night by .38 Special      1     1         18  18   NaN   
4              Art For Arts Sake by 10cc      1     1          1   1   NaN   

   <1970  
0  False  
1   True  
2  False  
3  False  
4  False

In [152]:
df.apply(col_convert)

SongClean  ARTISTCLEAN  ReleaseYear  COMBINED  First  Year  PlayCount  \
0           NaN          NaN       1982.0       NaN    1.0   1.0       82.0   
1           NaN          NaN          NaN       NaN    1.0   0.0        3.0   
2           NaN          NaN       1981.0       NaN    1.0   1.0       85.0   
3           NaN          NaN       1980.0       NaN    1.0   1.0       18.0   
4           NaN          NaN       1975.0       NaN    1.0   1.0        1.0   
5           NaN          NaN       2000.0       NaN    1.0   1.0       13.0   
6           NaN          NaN       2000.0       NaN    1.0   1.0        1.0   
7           NaN          NaN       2002.0       NaN    1.0   1.0        6.0   
8           NaN          NaN       1992.0       NaN    1.0   1.0        3.0   
9           NaN          NaN       1985.0       NaN    1.0   1.0        1.0   
10          NaN          NaN          NaN       NaN    1.0   0.0        1.0   
11          NaN          NaN       1980.0       NaN    1.0   1.0       97.0   
12          NaN          NaN       1993.0       NaN    1.0   1.0        6.0   
13          NaN          NaN          NaN       NaN    1.0   0.0        5.0   
14          NaN          NaN       1976.0       NaN    1.0   1.0       85.0   
15          NaN          NaN       1981.0       NaN    1.0   1.0       46.0   
16          NaN          NaN          NaN       NaN    1.0   0.0       24.0   
17          NaN          NaN       1995.0       NaN    1.0   1.0        1.0   
18          NaN          NaN       1980.0       NaN    1.0   1.0       39.0   
19          NaN          NaN       1980.0       NaN    1.0   1.0       74.0   
20          NaN          NaN       1979.0       NaN    1.0   1.0       92.0   
21          NaN          NaN       1975.0       NaN    1.0   1.0       39.0   
22          NaN          NaN       1984.0       NaN    1.0   1.0        1.0   
23          NaN          NaN       1977.0       NaN    1.0   1.0        3.0   
24          NaN          NaN          NaN       NaN    1.0   0.0        4.0   
25          NaN          NaN          NaN       NaN    1.0   0.0        2.0   
26          NaN          NaN          NaN       NaN    1.0   0.0       20.0   
27          NaN          NaN       1979.0       NaN    1.0   1.0        1.0   
28          NaN          NaN       1980.0       NaN    1.0   1.0       21.0   
29          NaN          NaN          NaN       NaN    1.0   0.0       45.0   
...         ...          ...          ...       ...    ...   ...        ...   
2200        NaN          NaN       1983.0       NaN    1.0   1.0       81.0   
2201        NaN          NaN       1971.0       NaN    1.0   1.0       44.0   
2202        NaN          NaN       1971.0       NaN    1.0   1.0        1.0   
2203        NaN          NaN          NaN       NaN    1.0   0.0        5.0   
2204        NaN          NaN       1971.0       NaN    1.0   1.0        1.0   
2205        NaN          NaN          NaN       NaN    1.0   0.0        2.0   
2206        NaN          NaN          NaN       NaN    1.0   0.0        1.0   
2207        NaN          NaN       1968.0       NaN    1.0   1.0       14.0   
2208        NaN          NaN          NaN       NaN    1.0   0.0        1.0   
2209        NaN          NaN       1976.0       NaN    1.0   1.0        1.0   
2210        NaN          NaN       1973.0       NaN    1.0   1.0        8.0   
2211        NaN          NaN       1980.0       NaN    1.0   1.0       34.0   
2212        NaN          NaN       1983.0       NaN    1.0   1.0       86.0   
2213        NaN          NaN       1983.0       NaN    1.0   1.0       29.0   
2214        NaN          NaN       1975.0       NaN    1.0   1.0        2.0   
2215        NaN          NaN          NaN       NaN    1.0   0.0       16.0   
2216        NaN          NaN          NaN       NaN    1.0   0.0       10.0   
2217        NaN          NaN       1973.0       NaN    1.0   1.0        6.0   
2218        NaN          NaN          NaN       NaN    1

#### 9.C Describe the new float-only DataFrame.

In [153]:
df.apply(col_convert).describe()

SongClean  ARTISTCLEAN  ReleaseYear  COMBINED   First         Year  \
count     2.000000          0.0  1653.000000       0.0  2230.0  2230.000000   
mean   1012.000000          NaN  1976.823351       NaN     1.0     0.741256   
std    1367.544515          NaN    54.330553       NaN     0.0     0.438043   
min      45.000000          NaN     0.000000       NaN     1.0     0.000000   
25%     528.500000          NaN  1971.000000       NaN     1.0     0.000000   
50%    1012.000000          NaN  1977.000000       NaN     1.0     1.000000   
75%    1495.500000          NaN  1984.000000       NaN     1.0     1.000000   
max    1979.000000          NaN  2014.000000       NaN     1.0     1.000000   

         PlayCount           FG  
count  2230.000000  2230.000000  
mean     16.872646    15.048430  
std      25.302972    25.288366  
min       0.000000     0.000000  
25%       1.000000     0.000000  
50%       4.000000     3.000000  
75%      21.000000    18.000000  
max     142.000000   142.000000